src/litserve/specs/openai_embedding.py

```diff
@@ -0,0 +1,155 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import asyncio
+import inspect
+import logging
+import time
+import uuid
+from typing import List, Literal, Optional, Union
+
+from fastapi import Request, Response, status
+from pydantic import BaseModel
+
+from litserve.specs.base import LitSpec
+from litserve.utils import LitAPIStatus
+
+logger = logging.getLogger(__name__)
+
+
+class EmbeddingRequest(BaseModel):
+    input: Union[str, List[str]]
+    model: str
+    dimensions: Optional[int] = None
+    encoding_format: Literal["float"] = "float"
+
+    # TODO: Check if this might be a handy helper function
+    def get_input_as_list(self):
+        return self.input if isinstance(self.input, list) else [self.input]
+
+
+class Embedding(BaseModel):
+    index: int
+    embedding: List[float]
+    object: Literal["embedding"] = "embedding"
+
+
+class UsageInfo(BaseModel):
+    prompt_tokens: int = 0
+    total_tokens: int = 0
+
+
+class EmbeddingResponse(BaseModel):
+    data: List[Embedding]
+    model: str
+    object: Literal["list"] = "list"
+    usage: UsageInfo
+
+
+LITAPI_VALIDATION_MSG = """LitAPI.predict and LitAPI.encode_response must not be a generator\
+(use return instead of yield) while using the OpenAI embedding spec.
+
+Error: {}
+
+Please follow the below example for guidance on how to use the spec:
+
+```python
+import numpy as np
+from typing import List
+from litserve.api import LitAPI
+
+
+class TestAPI(LitAPI):
+    def setup(self, device):
+        self.model = None
+
+    def decode_request(self, request)->List[str]:
+        return request.input if isinstance(request.input, list) else [request.input]
+
+    def predict(self, x)-> List[List[float]]:
+        return np.random.rand(len(x), 768).tolist()
+
+    def encode_response(self, output)-> dict:
+        return {"embeddings": output}
+```
+"""
+
+RESPONSE_VALIDATION_MSG = LITAPI_VALIDATION_MSG.split("\n", 2)[-1]
+
+
+class OpenAIEmbeddingSpec(LitSpec):
+    def __init__(self):
+        super().__init__()
+        # register the endpoint
+        self.add_endpoint("/v1/embeddings", self.embeddings, ["POST"])
+        self.add_endpoint("/v1/embeddings", self.options_embeddings, ["GET"])
+
+    def setup(self, server: "LitServer"):  # noqa: F821
+        from litserve import LitAPI
+
+        super().setup(server)
+
+        lit_api = self._server.lit_api
+        if inspect.isgeneratorfunction(lit_api.predict):
+            raise ValueError(LITAPI_VALIDATION_MSG.format("predict is a generator"))
```

Comment:
The error message needs to be more friendly.
Consider using the pattern: "What's wrong? Why it's wrong? How can I solve it?" when crafting error messages.

In this case it could be something like "You are using `yield` in your `predict` method, which is used for streaming. OpenAIEmbeddingSpec doesn't support streaming because producing embeddings is not a sequential operation. Please consider replacing `yield` with `return` in `predict`."

Reply:
Thank you for the clarification, @lantiga. This is very helpful! I'll make a note of this approach for crafting error messages and keep it as a reference for my other projects as well. Appreciate your guidance